packages/adblocker/src/engine/engine.ts

```diff
@@ -1498,13 +1525,80 @@ export default class FilterEngine extends EventEmitter<EngineEventHandlers> {
           result.exception = this.exceptions.match(request, this.isFilterExcluded.bind(this));
         }
+
+        // If the request is allowed in any way, we match request substitudes.
+        if (result.filter === undefined || result.exception !== undefined) {
+          const substitutions: Map<NetworkFilter, NetworkFilter | undefined> = new Map();
+          const url = new URL(request.url);
+          const substitudeFilters = this.substitudes.matchAll(
+            request,
+            this.isFilterExcluded.bind(this),
+          );
+          let modified = false;
+
+          // Handle $removeparam filters:
+          // 1st priority: @@||<entity>$removeparam
+          // 2nd priority: ||<entity>$removeparam
+          // 3rd priority: @@||<entity>$removeparam=x
+          // 4th priority: ||<entity>$removeparam=x
+          const removeparams: Map<string, NetworkFilter> = new Map();
+          const removeparamExceptions: Map<string, NetworkFilter> = new Map();
+          for (const filter of substitudeFilters) {
+            if (filter.isException()) {
+              removeparamExceptions.set(filter.removeparam!, filter);
+            } else {
+              removeparams.set(filter.removeparam!, filter);
+            }
+          }
+          const removeparamIgnoreFilter: NetworkFilter | undefined = removeparamExceptions.get('');
+          if (removeparams.size > 0 && url.searchParams.size > 0) {
+            for (const [key, filter] of removeparams) {
+              // Remove all params in case of option value is empty
+              // We will not match individual exceptions since it has a higher priority than them
+              if (key === '') {
+                substitutions.set(filter, removeparamIgnoreFilter);
+                // In case of non-existence of global exception, we will remove params
+                if (removeparamIgnoreFilter === undefined) {
+                  modified = true;
+                  url.search = '';
+                }
+                break;
+              }
+              if (!url.searchParams.has(key) && !key.startsWith('~')) {
+                continue;
+              }
+              const exception = removeparamExceptions.get(key) ?? removeparamIgnoreFilter;
+              substitutions.set(filter, exception);
+              if (exception === undefined) {
+                // Handle removeparam inversions
+                if (key.startsWith('~')) {
+                  const inversion = key.slice(1);
+                  for (const param of url.searchParams.keys()) {
+                    if (param !== inversion && !removeparamExceptions.has(param)) {
+                      modified = true;
+                      url.searchParams.delete(param);
+                    }
+                  }
+                } else {
+                  modified = true;
+                  url.searchParams.delete(key);
+                }
+              }
+            }
+          }
+          result.substitude = {
```

Comment:
In case there was a redirect + exception before, should we do something about `result.filter` and `result.exception`?

Reply:
Updated in 22cd3d71075ce5279916e6b88f56fbbde657beee but not sure if it matches your thought. Please, check the following section: [`22cd3d7` (#4528)](https://github.com/ghostery/adblocker/pull/4528/commits/22cd3d71075ce5279916e6b88f56fbbde657beee#diff-0a5765da7fa865c0ab71fc7b0187e3c6663bcc602558022d04dd7da3c0ab410aR1540-R1549)

Reply:
It'll directly impact to `result.matched`, so I feel it's reasonable to set values again once the lib finds a match in `removeparam`. However, the question is what to put in `result.filter` which only accepts one filter object.